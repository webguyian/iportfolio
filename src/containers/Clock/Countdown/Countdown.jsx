import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import Text from 'components/Text/Text';

import { useRadialOffset } from 'modules/clock/hooks';

const Countdown = props => {
  const { hours, minutes, seconds, radius, running } = props;
  const [offset, initialSeconds] = useRadialOffset(props, radius);
  const displaySeconds = seconds < 10 ? `0${seconds}` : seconds;
  const displayMinutes = minutes < 10 ? `0${minutes}` : minutes;
  const displayHours = hours > 0 ? `${hours}:` : '';
  const animationDuration = `${initialSeconds}s`;
  const animationPlayState = running ? 'running' : 'paused';
  const baseClass = 'countdown';
  const textClass = `${baseClass}-text`;
  const textWithHours = displayHours && `${textClass}--with-hours`;
  const style = {
    animationDuration,
    animationPlayState,
    strokeDashoffset: offset
  };

  return (
    <div className={baseClass}>
      <svg className={`${baseClass}-svg`}>
        <circle style={style} r={radius} cx={radius} cy={radius}></circle>
      </svg>
      <Text element="h2" className={cx(textClass, textWithHours)}>
        {`${displayHours}${displayMinutes}:${displaySeconds}`}
      </Text>
    </div>
  );
};

Countdown.propTypes = {
  allSeconds: PropTypes.number.isRequired,
  duration: PropTypes.number.isRequired,
  hours: PropTypes.number.isRequired,
  minutes: PropTypes.number.isRequired,
  radius: PropTypes.number,
  running: PropTypes.bool,
  seconds: PropTypes.number.isRequired
};

Countdown.defaultProps = {
  radius: 175,
  running: false
};

export default Countdown;
