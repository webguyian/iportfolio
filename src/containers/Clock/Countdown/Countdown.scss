$pi: 22/7;
$radius: 175;
$circumference: 2 * $pi * $radius;
$circle-size: #{$radius * 2}px;
$stroke-width: 8px;

.countdown {
  display: flex;
  position: relative;
  flex-shrink: 0;
  align-items: center;
  justify-content: center;
  width: 358px;
  height: 358px;
  border: $stroke-width solid $gray-dark;
  border-radius: $border-radius-round;

  &-text {
    @include type-number-display(8vw);
    margin: 0;

    &--with-hours {
      font-size: 7vw;
    }
  }

  &-svg {
    position: absolute;
    top: -5px;
    left: 50%;
    width: $circle-size;
    height: $circle-size;
    transform: translateX(-50%) rotateY(-180deg) rotateZ(-90deg);
    overflow: visible;

    > circle {
      stroke-dasharray: $circumference;
      stroke-dashoffset: 0;
      stroke-linecap: round;
      stroke-width: $stroke-width;
      stroke: $orange;
      fill: none;
      animation: countdown 10s linear infinite forwards;
    }
  }
}

@keyframes countdown {
  to {
    stroke-dashoffset: $circumference;
  }
}
