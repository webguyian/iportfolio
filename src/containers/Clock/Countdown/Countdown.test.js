import React from 'react';
import { create } from 'react-test-renderer';

import Countdown from './Countdown';

describe('<Countdown />', () => {
  const props = {
    allSeconds: 180,
    duration: 210,
    hours: 0,
    minutes: 3,
    seconds: 0
  };

  it('renders correctly', () => {
    const component = create(<Countdown {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with hours', () => {
    const updatedProps = {
      allSeconds: 4830,
      duration: 4830,
      hours: 1,
      minutes: 20,
      seconds: 30,
      running: true
    };

    const component = create(<Countdown {...updatedProps} />);

    expect(component).toMatchSnapshot();
  });
});
