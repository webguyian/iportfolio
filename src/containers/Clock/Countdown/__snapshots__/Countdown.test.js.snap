// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> renders correctly 1`] = `
<div
  className="countdown"
>
  <svg
    className="countdown-svg"
  >
    <circle
      cx={175}
      cy={175}
      r={175}
      style={
        Object {
          "animationDuration": "180s",
          "animationPlayState": "paused",
          "strokeDashoffset": 0,
        }
      }
    />
  </svg>
  <h2
    className="ui-text ui-text--body countdown-text"
  >
    03:00
  </h2>
</div>
`;

exports[`<Countdown /> renders correctly with hours 1`] = `
<div
  className="countdown"
>
  <svg
    className="countdown-svg"
  >
    <circle
      cx={175}
      cy={175}
      r={175}
      style={
        Object {
          "animationDuration": "4830s",
          "animationPlayState": "running",
          "strokeDashoffset": 0,
        }
      }
    />
  </svg>
  <h2
    className="ui-text ui-text--body countdown-text countdown-text--with-hours"
  >
    1:20:30
  </h2>
</div>
`;
